src/client/fuse_ll.cc

```diff
@@ -163,6 +167,66 @@ class CephFuse::Handle {
   struct fuse_args args;
 };
 
+static int already_fuse_mounted(const char *path, bool &already_mounted)
+{
+  struct statx path_statx;
+  struct statx parent_statx;
+  char *path_copy = NULL;
+  char *parent_path = NULL;
+  int err = 0;
+
+  already_mounted = false;
+
+  path_copy = strdup(path);
+  parent_path = dirname(path_copy);
+
+  // get stat information for original path
+  if ( -1 == statx(AT_FDCWD, path, AT_STATX_DONT_SYNC, STATX_INO, &path_statx) ) {
+    err = errno;
+    derr << "fuse_ll: already_fuse_mounted: statx(" << path << ") failed with error "
+      << cpp_strerror(err) << dendl;
+    goto out;
+  }
+
+  // if path isn't directory, then it can't be a mountpoint.
+  if ( !(path_statx.stx_mode & S_IFDIR) ) {
+    derr << "fuse_ll: already_fuse_mounted: "
+      << path << " is not a directory" << dendl;
+    goto out;
+  }
+
+  // get stat information for parent path
+  err = statx(AT_FDCWD, parent_path, AT_STATX_DONT_SYNC, STATX_INO, &parent_statx);
+  if ( -1 == statx(AT_FDCWD, parent_path, AT_STATX_DONT_SYNC, STATX_INO, &parent_statx) ) {
+    err = errno;
+    derr << "fuse_ll: already_fuse_mounted: statx(" << parent_path << ") failed with error "
+      << cpp_strerror(err) << dendl;
+    goto out;
+  }
+
+  // if original path and parent have different device ids,
+  // then the path is a mount point
+  // or, if they refer to the same path, then it's probably
+  // the root directory '/' and therefore path is a mountpoint
+  if( path_statx.stx_dev_major != parent_statx.stx_dev_major ||
+      path_statx.stx_dev_minor != parent_statx.stx_dev_minor ||
+      ( path_statx.stx_dev_major == parent_statx.stx_dev_major &&
+	path_statx.stx_dev_minor == parent_statx.stx_dev_minor &&
+	path_statx.stx_ino == parent_statx.stx_ino
+      )
+    ) {
+    // if getxattr returns positive length means value exist for ceph.client_id
+    // then ceph fuse is already mounted on path
+    if ( getxattr(path, _CEPH_CLIENT_ID, NULL, 0) > 0) {
+      already_mounted = true;
```

Comment:
output the client ID with `derr`.

Reply:
He's just getting the length here now and not worrying about the contents, which will tell us whether the thing is a ceph mount, but not whether this is the same ceph mount we would be doing. Is the value of the client_id useful here?

Also, we should note that this will prevent you from overlaying a kclient mount with a ceph-fuse mount, since it supports that xattr as well. I don't have a real issue with that, but it's not expected behavior.

Reply:
Actually I take it back -- we might as well do this right the first time.

@nmshelke : can you add in a statfs() check too? Do that and see if the f_type is FUSE. If it's not then you can return false.

Reply:
Sure